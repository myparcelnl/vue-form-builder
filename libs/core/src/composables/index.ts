export * from './useElementHooks';
export * from './useFormBuilder';
export * from './useLifecycleHooks';
export * from './useTestAttributes';
