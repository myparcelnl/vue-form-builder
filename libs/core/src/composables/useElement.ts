import {inject} from 'vue';
import {type AnyElementInstance} from '../types';
import {INJECT_ELEMENT} from '../symbols';

export const useElement = <I extends AnyElementInstance>(): I => {
  const element = inject(INJECT_ELEMENT) as I | undefined;

  if (!element) {
    throw new Error('No element found.');
  }

  return element;
};
