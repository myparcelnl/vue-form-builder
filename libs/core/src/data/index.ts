export * from './hooks';
export * from './provides';
