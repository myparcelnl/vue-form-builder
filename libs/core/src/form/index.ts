export * from './Form';
export * from './Form.types';
export * from './defaultFormConfiguration';
export * from './defineField';
export * from './defineForm';
export * from './interactive-element';
export * from './plain-element';
export * from './validator';
