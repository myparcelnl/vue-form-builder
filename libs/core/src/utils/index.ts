export * from './getValue';
export * from './markComponentAsRaw';
export * from './useDynamicWatcher';
