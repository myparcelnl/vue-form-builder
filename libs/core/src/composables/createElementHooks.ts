import {isUndefined, removePropertiesWith} from '@myparcel/ts-utils';
import {AnyElementInstance} from '../types';
import {HookCallback} from '@myparcel-vfb/hook-manager/src';
import {useLifecycleHooks} from './useLifecycleHooks';

type HookObject = Record<string, HookCallback | undefined>;

type UseElementHooks = (element: AnyElementInstance, initialHooks?: HookObject) => HookObject;

export const createElementHooks: UseElementHooks = (element, initialHooks = {}) => {
  const lifecycleHooks = useLifecycleHooks();

  lifecycleHooks.register(element.hooks, element);

  return removePropertiesWith(initialHooks, isUndefined);
};
