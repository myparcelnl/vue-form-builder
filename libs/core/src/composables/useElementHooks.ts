import {AnyElementInstance} from '../types';
import {HookCallback} from '@myparcel-vfb/hook-manager/src';
import {useLifecycleHooks} from './useLifecycleHooks';

type HookObject = Record<string, HookCallback | undefined>;

type UseElementHooks = (
  element: AnyElementInstance,
  initialHooks?: HookObject,
) => {
  hooks: HookObject;
};

export const useElementHooks: UseElementHooks = (element, initialHooks) => {
  const lifecycleHooks = useLifecycleHooks();

  lifecycleHooks.register(element.hooks, element);

  const registeredHooks = element.hooks.getRegisteredHooks();

  return {
    hooks: registeredHooks.reduce(
      (acc, hook) => ({
        ...acc,
        [hook.name]: hook.callback,
      }),
      initialHooks ?? {},
    ),
  };
};
