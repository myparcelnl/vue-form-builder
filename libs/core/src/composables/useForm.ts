import {inject} from 'vue';
import {type FormInstance} from '../form';
import {INJECT_FORM} from '../data';

export const useForm = (): FormInstance => {
  const form = inject(INJECT_FORM);

  if (!form) {
    throw new Error('No form found.');
  }

  return form;
};
