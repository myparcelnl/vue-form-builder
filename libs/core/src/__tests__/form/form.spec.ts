import {h, ref, reactive} from 'vue';
import {afterEach, describe, expect, it, vi} from 'vitest';
import {flushPromises} from '@vue/test-utils';
import {createField} from '../../utils';
import {Field} from '../../form';
import {getDefaultFormConfiguration} from '../../data';
import {useFormBuilder} from '../../composables';
import {renderTestForm} from './renderTestForm';
import {mockComponent} from './mockComponent';
import {getCommonFields} from './getCommonFields';

describe('rendering a form', () => {
  const formBuilder = useFormBuilder();

  afterEach(() => {
    formBuilder.setDefaults(getDefaultFormConfiguration());
    formBuilder.forms.value = Object.create(null);
  });

  it('renders html elements', async () => {
    const {wrapper} = await renderTestForm();
    const formElement = wrapper.find('form');

    expect(formElement.exists()).toBe(true);

    expect(wrapper.html()).toMatchSnapshot();

    expect(formElement.attributes('id')).toBe('test');
    expect(formElement.find('input').exists()).toBe(true);
    expect(formElement.findComponent(mockComponent).exists()).toBe(true);
  });

  it('adds class to form', async () => {
    const {wrapper} = await renderTestForm({
      form: {
        attributes: {
          class: 'form-class',
          'data-test': 'test',
        },
      },
    });
    const formElement = wrapper.find('form');

    expect(formElement.exists()).toBe(true);
    expect(formElement.classes()).toContain('form-class');
    expect(formElement.attributes('data-test')).toBe('test');
  });

  it('has form as default outer wrapper', async () => {
    const {wrapper} = await renderTestForm();
    const firstElement = wrapper.find('*');

    expect(firstElement.element.tagName).toBe('FORM');
  });

  it('has customizable outer wrapper tag', async () => {
    const {wrapper} = await renderTestForm({
      form: {
        tag: 'div',
      },
    });
    const firstElement = wrapper.find('*');

    expect(firstElement.element.tagName).toBe('DIV');
  });

  it('wraps fields with a wrapper element', async () => {
    const {wrapper} = await renderTestForm(
      {
        field: {
          wrapper: h('span', {class: 'inner-wrapper'}),
        },
      },
      [
        createField({
          name: 'field',
          component: mockComponent,
          ref: ref(''),
        }),
      ],
    );

    const wrapperElements = wrapper.findAll('span.inner-wrapper');

    expect(wrapperElements).toHaveLength(1);
    expect(wrapper.element).toMatchSnapshot();
  });

  it('adds attributes to each field', async () => {
    formBuilder.defaults.value.fieldDefaults.attributes.class = 'default-field-class';
    formBuilder.defaults.value.fieldDefaults.attributes['aria-label'] = 'test';

    const {wrapper} = await renderTestForm();
    const formElement = wrapper.find('form');

    expect(formElement.findAll('.default-field-class')).toHaveLength(3);
    expect(formElement.findAll('[aria-label="test"]')).toHaveLength(3);
  });

  it('makes fields lazy by default', async () => {
    const {form} = await renderTestForm({fieldDefaults: {lazy: true}});

    expect(form.instance.model.field1.lazy).toBe(true);
  });

  it('makes fields optional by default', async () => {
    const {form} = await renderTestForm({fieldDefaults: {optional: true}});

    expect(form.instance.model.field1.isOptional.value).toBe(true);
  });

  it('calls renderLabel if defined when rendering a label', async () => {
    expect.assertions(1);
    const renderLabel = vi.fn((value: string): string => value.toUpperCase());

    await renderTestForm({renderLabel}, getCommonFields());

    expect(renderLabel).toHaveBeenCalledTimes(3);
  });

  it('renders element prop if it is turned on', async () => {
    const {wrapper} = await renderTestForm();

    const input = wrapper.findComponent(mockComponent);

    expect(input.props('element')).toBeInstanceOf(Field);
  });

  it('does not render element prop if it is turned off', async () => {
    const {wrapper} = await renderTestForm({field: {elementProp: false}}, getCommonFields());

    const input = wrapper.findComponent(mockComponent);

    expect(input.props('element')).toBeUndefined();
  });

  describe('setValue', () => {
    it('sets the value of a single field', async () => {
      expect.assertions(1);
      const {form} = await renderTestForm();
      form.instance.setValue('field', '12345');
      await flushPromises();

      expect(reactive(form.instance.values)).toEqual({field: '12345', field2: ''});
    });

    it('does nothing if field does not exist', async () => {
      expect.assertions(1);
      const {form} = await renderTestForm();
      form.instance.setValue('does-not-exist', 'boo');
      await flushPromises();

      expect(reactive(form.instance.values)).toEqual({field: '', field2: ''});
    });
  });

  describe('setValues', () => {
    it('sets the value of multiple fields', async () => {
      expect.assertions(1);
      const {form} = await renderTestForm();
      form.instance.setValues({field1: 'bye', field2: 'hello'});
      await flushPromises();

      expect(reactive(form.instance.values)).toEqual({field1: 'bye', field2: 'hello'});
    });

    it('ignores fields that do not exist', async () => {
      expect.assertions(1);
      const {form} = await renderTestForm();
      form.instance.setValues({field1: 'value', wee: '3', field2: 'hi'});
      await flushPromises();

      expect(form.instance.values).toEqual({field1: 'value', field2: 'hi', field3: ''});
    });
  });
});
