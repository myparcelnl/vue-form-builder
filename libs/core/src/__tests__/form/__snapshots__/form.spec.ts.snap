// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`rendering a form > defaults to no inner wrapper 1`] = `
<form
  id="form"
>
  
  
  <div
    data-test-id="field"
    data-test-type="interactive"
  />
  <input
    data-test-id="named"
    data-test-type="plain"
  >
    
    
  </input>
  <br
    data-test-type="plain"
  >
    
    
  </br>
  
  <!--v-if-->
  
</form>
`;

exports[`rendering a form > renders html elements 1`] = `
"<form id=\\"form\\">
  <div data-test-id=\\"field\\" data-test-type=\\"interactive\\"></div><input data-test-id=\\"named\\" data-test-type=\\"plain\\"><br data-test-type=\\"plain\\">
  <!--v-if-->
</form>"
`;

exports[`rendering a form > wraps fields with a wrapper element 1`] = `
<form
  id="form"
>
  
  
  <span
    class="inner-wrapper"
    element="[object Object]"
  />
  <span
    class="inner-wrapper"
    element="[object Object]"
  />
  <span
    class="inner-wrapper"
    element="[object Object]"
  />
  
  <!--v-if-->
  
</form>
`;
